onnxscript/function_libs/torch_lib/ops/prims.py

```diff
@@ -176,12 +176,39 @@ def prims_bitwise_xor(self: TensorType, other: TensorType) -> TensorType:
     raise NotImplementedError()
 
 
+@torch_op("prims::broadcast_in_dim", trace_only=True)
 def prims_broadcast_in_dim(
     a: TensorType, shape: INT64, broadcast_dimensions: Sequence[int]
 ) -> TensorType:
     """broadcast_in_dim(Tensor(a) a, SymInt[] shape, int[] broadcast_dimensions) -> Tensor(a)"""
-
-    raise NotImplementedError()
+    
+    # Get the shape of the input tensor
+    input_shape = op.Shape(a)
+    target_rank = op.Size(shape)
+    
+    # Create the intermediate shape by constructing it with the right dimensions
+    # Start with a shape of all 1s
+    ones = op.ConstantOfShape(op.Unsqueeze(target_rank, axes=[0]), value=op.Constant(value_int=1))
+    
+    # Since broadcast_dimensions is known at compile time, we can create the mapping directly
+    # Convert broadcast_dimensions and input shape to tensors we can work with
+    broadcast_dims_tensor = op.Constant(value_ints=list(broadcast_dimensions))
+    input_rank = op.Size(input_shape)
+    indices = op.Range(op.Constant(value_int=0), input_rank, op.Constant(value_int=1))
+    
+    # Scatter the input dimensions into the intermediate shape at the specified positions
+    intermediate_shape = op.ScatterElements(
+        ones, 
+        op.Unsqueeze(broadcast_dims_tensor, axes=[0]), 
+        op.Unsqueeze(op.Gather(input_shape, indices), axes=[0]), 
+        axis=0
+    )
+    
+    # Reshape the input tensor to the intermediate shape
+    reshaped = op.Reshape(a, intermediate_shape)
+    
+    # Expand to the target shape  
+    return op.Expand(reshaped, shape)
```

Comment:
Is there a way to simplify and avoid op.ScatterElements, and Gather, which are costly?